### PR TITLE
[CS2113T-T12-3] OrgaNice!

### PR DESCRIPTION
**Pull request by terrytay**
@NizarMohd
@GanapathySanathBalaji
@hongquan448

### CODE REVIEW

File: text-ui-test/ACTUAL.TXT [hunk 1]
```diff
@@ -0,0 +1,47 @@
+Hello from
```
Comment (damithc) [inline text-ui-test/ACTUAL.TXT]: ACTUAL.TXT file should not be in git :-p

Comment (GanapathySanathBalaji) [reply text-ui-test/ACTUAL.TXT]: Thank you for the information Prof. We have stopped tracking ACTUAL.TXT in git.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,330 @@
...
+1. Task - Component contains details about handling the task list and related operations.
+
+1. StudyArea - Component contains details about handling queries for study area search.
+
+1. Exception - Component contains the various types of exceptions encountered when OrgaNice! is run.
+
+1. ResourceLoader - Component handles loading and saving of the task list and study area details to local storage.
+
+
+### 2.2. Task Component
```
Comment (jichngan) [inline docs/DeveloperGuide.md]: Good use of architecture diagram to illustrate software architecture.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,330 @@
...
+ The requirements captured are stored in the SchedulableTask object. 
+ Then, the TaskScheduler object finds the optimum schedule based on the user's requirements using the EDF 
+ (Early Deadline First) algorithm.
+    Since, EDF is an optimum algorithm, if it can't find a valid schedule it means that it's impossible to find a valid schedule based on the user's requirement. If a feasible schedule is found it is 
+ displayed, else a message stating that a schedule based on the user's requirements can't be made is displayed.
+ 
+	
+   The following sequence diagrams explain how tasks are scheduled.
+	
+ ![Overall Sequence Diagram](images/Schedule_Overall.png)
```
Comment (jichngan) [inline docs/DeveloperGuide.md]: You might want to explain what is the "ref" frame since it is not included in our syllabus.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,330 @@
...
 
-## Instructions for Manual Testing
+## 2. Design   
+ {Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}    
 
-{Give instructions on how to do a manual product testing e.g., how to load sample data to be used for testing}
+[comment]: # (@@author GanapathySanathBalaji)
+### 2.1. Architecture  
+  
+![Architecture](images/Architecture.png)
```
Comment (jichngan) [inline docs/DeveloperGuide.md:118]: Good use of Architecture diagram to get an overview of the Software Architecture.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,330 @@
...
+   2. StudyArea : Class that is used to model Study Areas. 
+   
+   3. StudyAreaList : Class that handles the list of available Study Areas based on User input.  
+        
+## 3. Implementation
+
+### 3.1. Scheduling Tasks 
+
+
+[comment]: # (@@author GanapathySanathBalaji)
```
Comment (jichngan) [inline docs/DeveloperGuide.md:232]: You might have to explain what the "ref" frame as it is not included in our syllabus at the moment.

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,330 @@
...
+ (Early Deadline First) algorithm.
+    Since, EDF is an optimum algorithm, if it can't find a valid schedule it means that it's impossible to find a valid schedule based on the user's requirement. If a feasible schedule is found it is 
+ displayed, else a message stating that a schedule based on the user's requirements can't be made is displayed.
+ 
+	
+   The following sequence diagrams explain how tasks are scheduled.
+	
+ ![Overall Sequence Diagram](images/Schedule_Overall.png)
+
+ The three reference frames used are as follows:
```
Comment (jichngan) [inline docs/DeveloperGuide.md:248]: You might want to check why is there another instance of an object at end of the lifeline of an object.

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,330 @@
...
+requirement is entered by the user, only the Study Areas appears in all the related attributes specified by the User 
+will be added to the output list.
+    - Pros : Data is categorised base on attributes
+    - Cons : More memory is required as more data structures are used. Since methods invoked during the search are also 
+    linear, time taken to complete the search will be longer if more data are present. Overall, this approach is harder 
+    to implement.
+  
+Therefore, the first alternative is chosen, as it is easier to implement and lesser memory is used while 
+conducting the search.
+
```
Comment (jichngan) [inline docs/DeveloperGuide.md:356]: Overall good usage of sequence diagrams to aid in the understanding of implementations. In the implementations section, it could be better if the steps during execution is listed out instead of writing the implementations in a long paragraph. Most of the sequence diagram has an object, which i assume is your main class, named as Duke. It could be better to either name it as main or your program name .

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,330 @@
...
+[comment]: # (@@author )
```
Comment (jichngan) [inline docs/DeveloperGuide.md]: Most of the sequence diagrams do not follow the standard notation of having just one instance of the class on the top. There is no need to mention the class again at the bottom. Please see the attached screenshot. 

Screenshot 2020-03-30 at 4 42 01 PM

**Review by jichngan — COMMENTED** [review]
Please look at the review as given in the DeveloperGuide.md document.

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,426 @@
...
+
+1. Task - Component contains details about handling the task list and related operations.
+
+1. StudyArea - Component contains details about handling queries for study area search.
+
+1. Notes - Component contains details about Notes related operations.
+
+1. Exception - Component contains the various types of exceptions encountered when OrgaNice! is run.
+
+1. ResourceLoader - Component handles loading and saving of the task list and study area details to local storage.
```
Comment (Bencotti) [inline docs/DeveloperGuide.md]: I really appreciate this short portion, at a quick glance I'm able to understand what the core classes are purposed for. This will help new developers get up to speed fast, coupled together with the object diagram.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,426 @@
...
 
-* *glossary item* - Definition
+## 2. Design   
+ {Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}    
 
-## Instructions for Manual Testing
+[comment]: # (@@author GanapathySanathBalaji)
+### 2.1. Architecture  
+  
+![Architecture](images/Architecture.png)
```
Comment (Bencotti) [inline docs/DeveloperGuide.md:118]: Would it make it clearer if different classes used unique colours or arrows to make it easier to differentiate the associations?

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,426 @@
...
+    
+   2. StudyArea : Class that is used to model Study Areas. 
+   
+   3. StudyAreaList : Class that handles the list of available Study Areas based on User input.  
+
+
+[comment]: # (@@author terrytay)
+
+### 2.4. Notes Component
+![Notes Component](images/NotesComponent.png)
```
Comment (Bencotti) [inline docs/DeveloperGuide.md:209]: I felt the arrow leading from ModuleManager was quite confusing as the branch suddenly became curved, and almost overlaps the top arrow. This might make some look for a separate origin rather than thinking of a split. 

It might be helpful to describe what CommandStack is for as it's the only floating unlinked object.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,426 @@
...
+    the method will iterate through the flags array and check if the Study Area's attribute matches each flag stated in
+    flags. If isAvailStudyArea returns as true for all flags, the Study Area is then added to the output list, 
+    availStudyArea. 
+
+Below would be a sequence diagram to demonstrate how the search algorithm is operated.
+
+ * `User enters search key` 
+ 
+![Study Area Sequence_Diagram_Main](images/user_sL_interaction.png)  
+<div>Figure 9. Interaction between User and Study Area Search Interface</div>
```
Comment (Bencotti) [inline docs/DeveloperGuide.md:328]: Is the format for replying back to user correct? Since it is a response, it might be easier to standardise and use a dotted arrow instead as with the rest of the diagram, to avoid confusion. 

sc-1

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,426 @@
...
+    to implement.
+  
+Therefore, the first alternative is chosen, as it is easier to implement and lesser memory is used while 
+conducting the search.
+
+[comment]: # (@@author terrytay)
+
+### 3.3. Operation of Notes
+
+#### 3.3.1 Implementation
```
Comment (Bencotti) [inline docs/DeveloperGuide.md:408]: This sections was a little too wordy, it might be better to have simple diagrams to help envision the flow as with the diagrams at the start.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,426 @@
...
+ |v2.0|user|create a Module|add notes inside|
+ |v2.0|user|undo an added note|increase my efficiency|
+ |v2.0|user|redo a removed notes|increase my efficiency|
+ |v2.0|user|create a schedule based on requirements|customise my tasks accordingly|
+
+## Appendix C: Non-Functional Requirements    
+  {Give non-functional requirements}    
+    
+[comment]: # (@@author NizarMohd)
+## Appendix D: Glossary
```
Comment (Bencotti) [inline docs/DeveloperGuide.md:556]: Very detailed glossary, it might be easier to reference if it was used in a table instead, what do you think?

File: docs/DeveloperGuide.md (hunk 11 above)
Comment (Bencotti) [inline docs/DeveloperGuide.md:209]: sc-2


The wrong UML notation might have been used for this object diagram, such as the naming convention of each object

**Review by Bencotti — COMMENTED** [review]
Great job detailing the workings of your project!

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,426 @@
...
 
-## Instructions for Manual Testing
+[comment]: # (@@author GanapathySanathBalaji)
+### 2.1. Architecture  
+  
+![Architecture](images/Architecture.png)
+<div>Figure 1. Overall Architecture of OrgaNice!</div>
+<br>
 
-{Give instructions on how to do a manual product testing e.g., how to load sample data to be used for testing}
```
Comment (yuxianglim) [inline docs/DeveloperGuide.md:34]: This part helps a lot in understanding the architecture of the program and the different classes and their functions efficiently. Good job!

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (yuxianglim) [inline docs/DeveloperGuide.md:118]: Maybe could consider making the arrows overlaps each other lesser so that it will be much neater or like bencotti suggested using unique colors or arrows. but good job anyway!

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,34 +1,426 @@
...
+    - Cons : More memory is required as more data structures are used. Since methods invoked during the search are also 
+    linear, time taken to complete the search will be longer if more data are present. Overall, this approach is harder 
+    to implement.
+  
+Therefore, the first alternative is chosen, as it is easier to implement and lesser memory is used while 
+conducting the search.
+
+[comment]: # (@@author terrytay)
+
+### 3.3. Operation of Notes
```
Comment (yuxianglim) [inline docs/DeveloperGuide.md:406]: Perhaps can consider  adding some diagram or  tables to illustrate the points so that readers can follow and flow easily.

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,34 +1,426 @@
...
+ * *isSizeCapable()* - This method returns true if the maximum size of the Study Area is capable of containing the size staed
+ by user.
+ * *containsSearchKey()* - This method returns true if the Study Area's name,faculty or address contains the search key 
+ entered by the user. This method is invoked when under the [default](#interaction-within-organices-submodule) condition as search by name, address or locations 
+ does not require any flags, instead it utilises a loose search. 
+ * *ports_flag* - refers to "-p" flag
+ * *indoor_flag* - refers to "-i" flag
+ * *outdoor_flag* - refers to "-o" flag
+ * *size_flag* - refers to "-s" flag
+
```
Comment (yuxianglim) [inline docs/DeveloperGuide.md]: Perhaps can consider putting this into a table or make the same space indentation so that it is easier to refer to.

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (bennychanya) [inline docs/DeveloperGuide.md:118]: Clear general flow of program using Architecture diagram. Perhaps the arrow from StudyArea to Ui should be separated from the arrow from Command to StudyArea to avoid confusion?

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,34 +1,426 @@
...
+### Testing for Study Area Search
+
+#### Search by location, name, address 
+ * To test for accuracy of loose search, test "bux" to see if it returns locations related to Starbucks.
+#### Search by flags only 
+ * To test for accuracy of flags, test either "-p", "-i", "-o" or "-s {integer}"
+#### Search with both, (1) location, name or address , and , (2) flags 
+ * To test for accuracy, test "{location/name/address} {flags}".
+ * Since flags must come as a second argument in this case, test for "{flags} {location/name/address}"
```
Comment (yuxianglim) [inline docs/DeveloperGuide.md]: DG is very detailed and explains the classes well. Can consider doing a full rundown of the sequence of events with a diagram or numbered too. Goodjob:)

File: docs/DeveloperGuide.md (hunk 11 above)
Comment (bennychanya) [inline docs/DeveloperGuide.md:209]: Should the arrows be separated to clearly show the flow of the diagram?

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,34 +1,426 @@
...
+1. NotesInvoker : Class to start the Notes
+	
+2. Notes : Class to support operations for the modules in the Notes, acts as interface.
+	
+3. Modulelist: Class to implement actual modules operations and store modules list.
+	
+4. Command : Package containing Command interface, Add command, Command Stack classes.
+	
+5. Parser : Class to parse commands for command-based operations.
+
```
Comment (bennychanya) [inline docs/DeveloperGuide.md]: Perhaps there should be an explanation for the role of the ModuleManager?

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1,34 +1,426 @@
...
+ 1. The requirements captured are stored in the SchedulableTask object. 
+ 1. Then, the TaskScheduler object finds the optimum schedule based on the user's requirements using the EDF 
+ (Early Deadline First) algorithm.
+ 1. If a feasible schedule is found it is 
+ displayed, else a message stating that a schedule based on the user's requirements can't be made is displayed.
+ 
+	
+   The following sequence diagrams explain how tasks are scheduled.
+	
+ ![Overall Sequence Diagram](images/Schedule_Overall.png)
```
Comment (bennychanya) [inline docs/DeveloperGuide.md]: Should the word size of the diagram be increased or maybe the long commands could be split into 2 lines? It's quite hard to read the diagram without going closer to the screen.

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1,34 +1,426 @@
...
+- getFlagsInfo() : 
+    - This method will retrieve the flags entered by User, at the same time validating the User input. The criteria 
+    entered by the User will be returned as flags, an Array of Strings of fixed size 5 as the software only supports 5 
+    flags in version 2.
+- isAvailStudyArea() : 
+    - This method does the check mentioned above and returns a boolean value. Based on the flags detected, 
+    the method will iterate through the flags array and check if the Study Area's attribute matches each flag stated in
+    flags. If isAvailStudyArea returns as true for all flags, the Study Area is then added to the output list, 
+    availStudyArea. 
+
```
Comment (bennychanya) [inline docs/DeveloperGuide.md:322]: Maybe some of these descriptions should be placed between the diagrams for diagram explanation.

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -1,34 +1,426 @@
...
+ * `StudyAreaCommand invokes searchList() of StudyAreaList` 
+ 
+![Study_Area_Sequence_Diagram_subModules](images/sL_sA_interaction.png)
+<div>Figure 10. Interaction within Study Area Search Interface</div>
+<br>
+
+![Study_Area_Sequence_Diagram_subModules2](images/isAvailStudyArea.png)
+<div>Figure 11. Interaction when isAvailStudyArea is invoked</div>
+<br>
```
Comment (bennychanya) [inline docs/DeveloperGuide.md:341]: Clear diagram to showcase how the method works. Good job!
